collect_app/src/test/java/org/odk/collect/android/preferences/AdminPasswordDialogFragmentTest.java

```diff
@@ -0,0 +1,116 @@
+import org.odk.collect.android.support.TestActivityScenario;
+import org.odk.collect.android.utilities.AdminPasswordProvider;
+import org.robolectric.annotation.LooperMode;
+
+import static android.os.Looper.getMainLooper;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.odk.collect.android.preferences.AdminPasswordDialogFragment.ARG_ACTION;
+import static org.odk.collect.android.preferences.AdminPasswordDialogFragment.Action.STORAGE_MIGRATION;
+import static org.robolectric.Shadows.shadowOf;
+import static org.robolectric.annotation.LooperMode.Mode.PAUSED;
+
+@RunWith(AndroidJUnit4.class)
+@LooperMode(PAUSED)
+public class AdminPasswordDialogFragmentTest {
+
+    @Before
+    public void setup() {
+        RobolectricHelpers.overrideAppDependencyModule(new AppDependencyModule() {
+            @Override
+            public AdminPasswordProvider providesAdminPasswordProvider() {
+                return new StubAdminPasswordProvider();
+            }
+        });
+    }
+
+    @Test
+    public void enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = createFragment();
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
+        });
+    }
+
+    @Test
+    public void afterRecreating_enteringPassword_andClickingOK_callsOnCorrectAdminPasswordWithAction() {
+        TestActivityScenario<MockAdminPasswordDialogCallback> activityScenario = TestActivityScenario.launch(MockAdminPasswordDialogCallback.class);
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = createFragment();
+            fragment.show(activity.getSupportFragmentManager(), "tag");
+            shadowOf(getMainLooper()).idle();
+        });
+
+        activityScenario.recreate();
+        activityScenario.onActivity(activity -> {
+            AdminPasswordDialogFragment fragment = (AdminPasswordDialogFragment) activity.getSupportFragmentManager().findFragmentByTag("tag");
+
+            fragment.getInput().setText("password");
+            ((AlertDialog) fragment.getDialog()).getButton(AlertDialog.BUTTON_POSITIVE).performClick();
+            shadowOf(getMainLooper()).idle();
+
+            assertThat(activity.onCorrectAdminPasswordCalledWith, equalTo(STORAGE_MIGRATION));
+        });
+    }
+
```

Comment:
While you're here, would you consider that the correct callback method is called if the password is incorrect?

Reply:
Yeah sounds good